src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -19,63 +27,186 @@
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest == null || pullRequest.isEmpty()) {
-            return Optional.empty();
-        }
-        return pullRequest.stream().filter(pr -> pr.getId() == pullRequestId).findFirst();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        return Optional.ofNullable(pullRequests.getOrDefault(cacheKey, new RepositoryStore(new ConcurrentHashMap<>()))
+                .getPullRequest(fromBranch, toBranch));
     }
 
     @Override
-    public void refreshStore(String key, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
+    public void refreshStore(String projectKey, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest != null) {
-            pullRequest.clear();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        AtomicLong oldestUpdate = pullRequests.getOrDefault(cacheKey, new RepositoryStore(new ConcurrentHashMap<>())).findOldestUpdate();
+
+        //bbspullrequests is ordered from newest to oldest
+        bbsPullRequests.map(pr -> {
+            if (pr.getUpdatedDate() > oldestUpdate.get()) {
+                updatePullRequest(serverId, pr);
+            }
+            return pr.getUpdatedDate();
+        }).filter(updated -> updated < oldestUpdate.get())
+                .findFirst();
+    }
+
+    @Override
+    public void removeClosedPullRequests(long date) {
+       pullRequests.values().forEach(repositoryStore -> repositoryStore.removeClosedPullRequests(date));
+    }
+
+    @Override
+    public boolean hasPRForRepository(String projectKey, String slug, String serverId) {
+        PullRequestStoreImpl.CacheKey key =
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        return pullRequests.getOrDefault(key, new RepositoryStore(new ConcurrentHashMap<>())).hasPR();
+    }
+
+    public static MinimalPullRequest convert(BitbucketPullRequest bbsPR) {
+        return new MinimalPullRequest(bbsPR.getId(), bbsPR.getState(), bbsPR.getFromRef().getDisplayId(),
+                bbsPR.getToRef().getDisplayId(), bbsPR.getUpdatedDate());
+    }
+
+    private static class RepositoryStore {
+
+        private static final MinimalPullRequest CLOSED_PR = new MinimalPullRequest(0, BitbucketPullState.DECLINED,
+                "", "", 0);
+        private ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests;
```

Comment:
I think a comment explaining what the key and values here are relative to a PR is necessary, to make this code more readable- using similar terminology you use for stream variable names (like `toMap`, `ref` and so on)